api/envoy/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto

```diff
@@ -70,4 +70,19 @@ message ProxyProtocol {
   //   :ref:`core.v3.ProxyProtocolConfig.pass_through_tlvs <envoy_v3_api_field_config.core.v3.ProxyProtocolConfig.pass_through_tlvs>`,
   //   which controls pass-through for the upstream.
   config.core.v3.ProxyProtocolPassThroughTLVs pass_through_tlvs = 3;
+
+  // The PROXY protocol versions to match. See https://www.haproxy.org/download/2.1/doc/proxy-protocol.txt for details.
+  // By default, the filter will match any version.
+  //
+  // When the filter receives PROXY protocol data that does not match the specified versions,
+  // it will reject the connection.
+  //
+  // .. attention::
+  //
+  //   When used in conjunction with the :ref:`allow_requests_without_proxy_protocol <envoy_v3_api_field_extensions.filters.listener.proxy_protocol.v3.ProxyProtocol.allow_requests_without_proxy_protocol>`,
+  //   the filter will only attempt to match signatures for the allowed versions.
+  //   For example, when ``allowed_versions=V1``, ``allow_requests_without_proxy_protocol=true``,
+  //   and an incoming request matches the V2 signature, the filter will allow the request through as if it did not contain
+  //   PROXY protocol information.
```

Comment:
Is it today's behavior when ``allow_requests_without_proxy_protocol=true`` and an incoming request matches the V2 signature but not contain PROXY protocol info?

Reply:
>  but not contain PROXY protocol info

Sorry, I do not understand the question fully. If the request has a PROXY protocol V2 signature, then the filter matches it to PROXY protocol V2 and `allow_requests_without_proxy_protocol` does not matter. If there are any errors when parsing the rest of the PROXY protocol V2 request, then it will be an error and rejected, otherwise request will be allowed through.

This change here mentions when `allowed_versions=V1` so the filter will *not* attempt to match the PROXY protocol v2 signature.

Reply:
I probably get confused on what does mean the v2 signature matched but did not contain the proxy protocol information. Then the question is it error case? or it is a case spec said it is ok to provide a v2 signature but without proxy protocol information.

Reply:
I think there is a misunderstanding in the documentation I added in this PR. The line states:

> an incoming request matches the V2 signature, the filter will allow the request through **as if** it did not contain PROXY protocol information.

What I mean here: When a valid PROXY protocol request with V2 signature hits the filter, because `allowed_versions=V1`, the filter will **not** attempt to match the request against PROXY protocol v2. It will treat it as a request without PROXY protocol. In this case, the filter will by default reject the request. If `allow_requests_without_proxy_protocol=true`, the filter will allow the request (filter treats it as a non-PROXY protocol request).

I don't think it has to do with the spec. It's just limiting the implementation of the filter. The filter is still spec compliant by default, but a user can choose to reduce which parts of the spec the filter matches.

Reply:
Rephrased the proto docs to make this clearer.

Reply:
thanks! It's much clearer